ssl/quic/quic_channel.c

```diff
@@ -1283,8 +1283,12 @@ static int ch_on_transport_params(const unsigned char *params,
                 goto malformed;
             }
 
-            /* Must match SCID of first Initial packet from server. */
-            if (!ossl_quic_conn_id_eq(&ch->init_scid, &cid)) {
+            /*
+             * Must match SCID of first Initial packet from server.
+             * Keep in mind the client always sends QUIC_TPARAM_INITIAL_SCID
+             * empty (cid.id_len == 0, see ch_generate_transport_params()).
+             */
+            if (cid.id_len != 0 && !ossl_quic_conn_id_eq(&ch->init_scid, &cid)) {
```

Comment:
Why would we get the INITIAL SCID transport parameter if the initial SCID was empty? I do  not understand this change, can you please explain?

Reply:
I think this is in the event that a retry cycle was done no?  I.e. the client sends an empty inital scid param, but if the server is doing a retry operation, it encodes the inital SCID as a transport param, which the client then needs to validate as matching the initial_scid it sent in the first packet.

Reply:
Why wouldn't the `ch->init_scid` be empty as well then?

Reply:
IIRC it gets assigned from the packet header src_conn_id in ch_rx_handle_packet when the retry frame gets sent

Reply:
This is very good question indeed. Took me a while to figure out what's going on here. The port for listener object gets created in multiconnection mode, snippet below comes from `ossl_quic_new_listener()`
```
4234     port_args.channel_ctx       = ctx;
4235     port_args.is_multi_conn     = 1;
4236     ql->port = ossl_quic_engine_create_port(ql->engine, &port_args);
```
the QUIC connection object created from listener by `SSL_new_from_listener()` shares the port with listener it comes from. The `multi_conn` property becomes `QUIC_PORT::is_multi_conn` property. It is used here in `port_init()`:
```
  93 static int port_init(QUIC_PORT *port)
  94 {
  95     size_t rx_short_dcid_len = (port->is_multi_conn ? INIT_DCID_LEN : 0);
  96
  97     if (port->engine == NULL || port->channel_ctx == NULL)
  98         goto err;
  99
 100     if ((port->err_state = OSSL_ERR_STATE_new()) == NULL)
 101         goto err;
....
 119                                                                                                                                                                                                                                                                                                                        
 120     port->rx_short_dcid_len = (unsigned char)rx_short_dcid_len;
 121     port->tx_init_dcid_len  = INIT_DCID_LEN;
 122     port->state             = QUIC_PORT_STATE_RUNNING;
```
So setting mult_conn to 1 to be rx_short_dcid_len nonzero. This causes `ch_init()` to generate non-zero length connection id right here:
```
 262
 263     if (!ch->is_server
 264         && !ossl_quic_lcidm_generate_initial(ch->lcidm, ch, &txp_args.cur_scid))
 265         goto err;
 266
 267     /* We use a zero-length SCID. */
 268     txp_args.cur_dcid               = ch->init_dcid;
 269     txp_args.ack_delay_exponent     = 3;
```
After seeing this my initial thought was it's bad and `SSL_new_from_listener()` should somehow fix-up this to keep `cur_scid` zero for newly created QUIC connection objects  Then I want back to RFC 9000 and skimmed through Sections 5 and Section 7. I think this entirely fine to let QUIC connection objects created by `SSL_new_from_listener()` to send non-zero SCIDs. We just need to take that into account in OpenSSL QUIC stack
as seen in changes to `ssl/quic/quic_channel.c` done here.

Reply:
It absolutely makes sense to have non zero length source CID for SSL_new_from_listener() connections. You would not be able to distinguish between the incoming packets on the listener otherwise.

However we could still keep using zero length source CIDs on SSL_new() connections. Not sure if it is worth the hassle though.

Reply:
QUIC connection objects created by `SSL_new()` still use zero length CIDs. There is no change here. Unlike `ossl_quic_new_listener()` the `ossl_quic_new()` does not set `is_multi_conn` property to port.

It's worth to note the difference between SSL_new() and SSL_new_from_listener() here. Each object created by SSL_new() gets its own port/engine. WHile objects created by SSL_new_from_listener() share the same port/engine instance which got created with listener. So I think it's quite natural connections which come from listener use non-zero SCID (source connection id), while connections which come from SSL_new use zero length SCID.